### PR TITLE
GPU efficient Densenets

### PR DESCRIPTION
**Pull request by ekagra-ranjan**
In reference to #691, this PR provides the option for memory efficient implement of densenet models. 

I tested the models (original implementation and the new implementation with `efficient=False` as well as with `efficient=True`) on [hymenoptera dataset](https://download.pytorch.org/tutorial/hymenoptera_data.zip) which gave the follow results

Benchmark results (Batch size = 8, image size = 224x224):
 ```

                                                  Time taken         GPU Memory Consumption
Original                                          2m 55s                    1668mb
New (`efficient=False`)                           2m 58s                    1667mb
New (`efficient=True`)                            4m 6s                     1115mb

```
There was no significant change in the accuracy of the trained models. The implementation does not change the performance in terms of accuracy.

The new implementation with `efficient=True` seems to comsume ~1.5 times lesser GPU memory at the cost of ~1.4 times increased compute time. 

cc: @soumith

### CODE REVIEW

**Comment by ekagra-ranjan** [comment]
@soumith Is the PR fine?

**Comment by ekagra-ranjan** [comment]
@soumith  Is there anything else that needs to be done?

File: torchvision/models/densenet.py [hunk 1]
```diff
@@ -28,21 +38,40 @@ def __init__(self, num_input_features, growth_rate, bn_size, drop_rate):
...
+                growth_rate=growth_rate,
+                bn_size=bn_size,
+                drop_rate=drop_rate,
+                efficient=efficient,
+            )
             self.add_module('denselayer%d' % (i + 1), layer)
 
+    def forward(self, init_features):
+        features = [init_features]
+        for name, layer in self.named_children():
```
Comment (soumith) [inline torchvision/models/densenet.py]: is the ordering of this correct across all python versions?

File: torchvision/models/densenet.py [hunk 2]
```diff
@@ -16,8 +17,17 @@
 }
 
 
+def _bn_function_factory(norm, relu, conv):
+    def bn_function(*inputs):
```
Comment (soumith) [inline torchvision/models/densenet.py]: i forgot why this concatenation is needed for checkpointing, can you remind me?

Comment (ekagra-ranjan) [reply torchvision/models/densenet.py]: In a densenet block, the previous outputs are concatenated with current input before passing through a layer. So the checkpoints saves memory by not saving them in the computation graph for backward pass. Instead it recomputes these intermediate activation during backward pass which makes them slower.

File: torchvision/models/densenet.py (hunk 1 above)
Comment (ekagra-ranjan) [reply torchvision/models/densenet.py]: Worked on python 3.6. Will check them on 2.7 as well.

Comment (ekagra-ranjan) [reply torchvision/models/densenet.py]: @soumith Works on python 2.7.

**Comment by soumith** [comment]
I re-reviewed it today as well. After the test above ^^ which makes sure the same function is computed, this PR is good to go.

**Comment by ekagra-ranjan** [comment]
I have added the test but there are conflicts. @fmassa Will you be able to resolve it?

**Comment by fmassa** [comment]
@ekagra-ranjan do you want me to resolve the conflicts?

**Comment by ekagra-ranjan** [comment]
Yes @fmassa, that would be very helpful.

**Comment by fmassa** [comment]
I've sent a new PR in https://github.com/pytorch/vision/pull/1003

All the history of changes that you have made have been kept.
Thanks a lot for the awesome work @ekagra-ranjan !
